syncronize camera_info with images

LeroyR opened the pull request:
### Description

camera_info needs to have the same timestamp as the images. (e.g. for usage with `message_filters.TimeSynchronizer`)

- give each image from cameras its own CameraInfo
- add configuration for camera topics
    
This changes cameras to publish a syncronized camera_info for each image. e.g for rgb image:
    `/mujoco_server/cameras/camera_name/rgb/camera_info`
    `/mujoco_server/cameras/camera_name/rgb/image_raw`
    
Additional configuration options to change topics:
      `camera_name/topic`: defaults to "/mujoco_server/cameras/camera_name"
      `camera_name/name_rgb`: default to "rbg"
      `camera_name/name_depth`: default to "depth"
      `camera_name/name_segment`: default to "segmented"

### Checklist
- [ ] **Required by CI**: Code is auto formatted using clang-format.
- [ ] Add a brief explanation of your changes to the [changelog](CHANGELOG.md).

LeroyR committed:
syncronize and split camera_info with images.

- give each image from cameras its own topic
- add configuration for camera topics

This changes cameras to publish a syncronized camera_info for
each image. e.g for rgb image:
/mujoco_server/cameras/camera_name/rgb/camera_info
/mujoco_server/cameras/camera_name/rgb/image_raw

Additional configuration options to change topics:
  camera_name/topic: defaults to "/mujoco_server/cameras/camera_name"
  camera_name/name_rgb: default to "rbg"
  camera_name/name_depth: default to "depth"
  camera_name/name_segment: default to "segmented"

DavidPL1 commented on mujoco_ros/src/offscreen_camera.cpp:
```diff
@@ -298,15 +301,15 @@ void OffscreenCamera::renderAndPublish(mujoco_ros::OffscreenRenderContext *offsc
 
-	if (rendered) {
-		publishCameraInfo();
+	if (!rendered) {
```
why is it necessary to publish camera info when no image was rendered?

LeroyR replied:
More like a HACK/QUICKFIX: my rationale was still sending `camera_info` even if nothing is subscribed to the image topics. 
Maybe the code should be restructured to send camera_info for each image topic separately and start rendering and publishing on either info or image topic subscription.

DavidPL1 commented on mujoco_ros/src/offscreen_camera.cpp:
```diff
@@ -250,6 +250,9 @@ bool OffscreenCamera::renderAndPubIfNecessary(mujoco_ros::OffscreenRenderContext
 		depth_pub_.publish(depth_msg);
 	}
+
+	publishCameraInfo(ros::Time(scn_state_.data.time));
```
renderAndPubIfNecessary can be called twice when a stream type contains both RGB and SEGMENT. Since they all share the same camera info manager, this means republishing the same data. In case it is necessary to publish camera info without any rendered images it makes more sense making the publish below (in renderAndPublish) unconditional.

LeroyR replied:
AFAIK all cameras publish info for each image separately. I could not find any definite answer if that is required anywhere beyond the information in [rep-0104](https://github.com/ros-infrastructure/rep/blob/master/rep-0104.rst#publish-camerainfo-only-when-parameters-change) and the [image_geometry](http://wiki.ros.org/image_geometry) package.

DavidPL1 commented on mujoco_ros/src/offscreen_camera.cpp:
```diff
@@ -298,15 +301,15 @@ void OffscreenCamera::renderAndPublish(mujoco_ros::OffscreenRenderContext *offsc
+void OffscreenCamera::publishCameraInfo(const ros::Time &t)
 {
 	sensor_msgs::CameraInfo camera_info_msg = camera_info_manager_->getCameraInfo();
-	camera_info_msg.header.stamp            = ros::Time::now();
```
Technically this shouldn't make a difference. The physics loop waits until no offscreen render request are pending, thus ros::Time(scn_state_.data.time) should be equal to ros::Time::now()

LeroyR replied:
I think you assumption does not hold:

Without the patch i get this:

```
rostopic echo /mujoco_server/cameras/camera/rgb | grep -B 1 nsecs:
--
    secs: 109
    nsecs: 276000000
--
    secs: 109
    nsecs: 344000000
--
    secs: 109
    nsecs: 480000000
--
    secs: 109
    nsecs: 684000000
```

```
rostopic echo /mujoco_server/cameras/camera/camera_info | grep -B 1 nsecs
--
    secs: 109
    nsecs: 210000000
--
    secs: 109
    nsecs: 278000000
--
    secs: 109
    nsecs: 346000000
--
    secs: 109
    nsecs: 414000000
--
    secs: 109
    nsecs: 482000000
--
    secs: 109
    nsecs: 550000000
--
    secs: 109
    nsecs: 618000000
--
    secs: 109
    nsecs: 686000000
--
    secs: 109
    nsecs: 754000000
--
    secs: 109
    nsecs: 822000000
--
    secs: 109
    nsecs: 890000000
--
    secs: 109
    nsecs: 958000000

```

DavidPL1 commented:
I think this PR should be extended to allow different resolutions per stream type (probably RGB and SEGMENT should use the same) , justifying the need of individual camera infos. To better replicate real setups.
Following [rep-0104](https://github.com/ros-infrastructure/rep/blob/master/rep-0104.rst) we then can keep the model of only publishing camera info when an image was rendered, but change the rendering condition to a subscription of either the Image or the info topic.
I suppose the information on how to align RGB and Depth needs to be specified, too, in case resolutions differ.

rhaschke commented:
> camera_info needs to have the same timestamp as the images

Are you sure. Usually, camera_info topics are latched and simply provide a fixed message. Therefore, the timestamp shouldn't be relevant.

LeroyR commented:
> > camera_info needs to have the same timestamp as the images
> 
> Are you sure. Usually, camera_info topics are latched and simply provide a fixed message. Therefore, the timestamp shouldn't be relevant.

Atleast our cameras do that. By [rep-0104](https://github.com/ros-infrastructure/rep/blob/master/rep-0104.rst) `camera_info` is send with every image:

> Publish CameraInfo Only When Parameters Change
One FAQ is, "Why send CameraInfo with every Image message? Why not only once?" The main answer is that operational parameters (binning, ROI) may change, perhaps rapidly, from image to image.

LeroyR commented:
I changed the offscreen_camera to have a `CameraInfo` publisher for each of the images.

The topics are now configurable such that:
```
/mujoco_server/cameras/camera_name/rgb/image_raw
/mujoco_server/cameras/camera_name/rgb/camera_info
```

is used as default - which should match real camera setups.

DavidPL1 committed:
test: makes BaseEnvFixutre cleanup params on teardown

DavidPL1 committed:
chg: headless no longer implies no offscreen

DavidPL1 committed:
fix: freeing offscreen context in headless mode

previously, in case a render request is still active, freeing the context
would case a segfault. Instead now the render mutex is locked, the
pending request is canceled and the context freed.

DavidPL1 committed:
build: adds option to build w/ GLFW, EGL or OSMesa

Default build tries the following:
If GLFW is found, it will be used for GUI and offscreen rendering.
Otherwise GUI functionality is not provided. As fallback for offscreen
rendering first GLE is searched, otherwise osmesa. In case neither can
be found, offscreen rendering is disabled, too.

DavidPL1 committed:
fix: offscreen render pipeline improvements

* fixes edge case where model was missing fields because of a missing
forward computation which caused issues when applying the model to the
offscreen scene.
* fixes possible deadlock during shutdown when waiting to join the render
loop.

DavidPL1 committed:
chg: updates BaseEnvFixture and tests using it

DavidPL1 committed:
test: adds initial offscreen rendering tests

DavidPL1 committed:
gh: adds build deps for EGL/OSMESA to ci docker

DavidPL1 committed:
gh: adds workflow combinations for render backend

DavidPL1 committed:
fix: segmentation image rendering

DavidPL1 committed:
fix: sets off-cam topic member correctly on init

DavidPL1 committed:
test: adapt render tests to additional params

DavidPL1 committed:
chg: render backend cmake changes

DavidPL1 committed:
tests: fix tests

DavidPL1 committed:
fix: corrects offscreen scene creation

DavidPL1 committed:
fix: adds bool to signal OSMesa cleanup necessity

DavidPL1 commented:
FYI: this is already merged into [wip-headless-rendering](https://github.com/ubi-agni/mujoco_ros_pkgs/tree/wip-headless-rendering). And will be available in noetic-devel once we merge the feature branch.

DavidPL1 commented:
This is now merged into noetic-devel. Apparently GitHub failed to notice
